### PR TITLE
Improve observability of enqueuing requests to query-scheduler

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR improves the observability of query-frontend interactions with query-schedulers.

In particular:

* tracing events are emitted when attempts are made to enqueue a request, when enqueuing fails and when it succeeds
* tracing events are emitted when a query response is received
* a span is emitted for each enqueue request to a scheduler, including the target scheduler's address
* the error returned by the scheduler, rather than `nil`, is now logged when the scheduler returns an error ([here](https://github.com/grafana/mimir/commit/e354aa417970b2577c3d5bb5ca9d12218e16e03f#diff-415639b83f694aaab7e3eb07f38b866ea9eac68ffa71407af7725a7dcb175296R409))

Example trace looks like this now:

Screenshot 2023-08-23 at 4 06 44 pm

Note to reviewers: I'd suggest reviewing each commit separately.

#### Which issue(s) this PR fixes or relates to

(none)

#### Checklist

- [n/a] Tests updated
- [n/a] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/frontend/v2/frontend.go [hunk 1]
```diff
@@ -222,9 +226,12 @@ func (f *Frontend) RoundTripGRPC(ctx context.Context, req *httpgrpc.HTTPRequest)
 	retries := f.cfg.WorkerConcurrency + 1 // To make sure we hit at least two different schedulers.
 
 enqueueAgain:
+	level.Debug(spanLogger).Log("msg", "enqueuing request")
```
Comment (charleskorn) [inline pkg/frontend/v2/frontend.go:229]: Note to reviewers: I've deliberately added this request started event and the "request succeeded" event below despite adding the span in `enqueueRequest`, as it's possible there's a delay between a request being added to `f.requestsCh` and it being picked up and enqueued by a worker.

File: pkg/frontend/v2/frontend_scheduler_worker.go [hunk 2]
```diff
@@ -408,3 +359,71 @@ func (w *frontendSchedulerWorker) schedulerLoop(loop schedulerpb.SchedulerForFro
...
+		QueryID:         req.queryID,
+		UserID:          req.userID,
+		HttpRequest:     req.request,
+		FrontendAddress: w.frontendAddr,
+		StatsEnabled:    req.statsEnabled,
+	})
+	w.enqueuedRequests.Inc()
+
+	if err != nil {
+		level.Warn(spanLogger).Log("msg", "received error while enqueuing request", "err", err)
```
Comment (pstibrany) [inline pkg/frontend/v2/frontend_scheduler_worker.go:380]: nit:

```suggestion
		level.Warn(spanLogger).Log("msg", "received error while sending request to scheduler", "err", err)
```

File: pkg/frontend/v2/frontend_scheduler_worker.go [hunk 3]
```diff
@@ -408,3 +359,71 @@ func (w *frontendSchedulerWorker) schedulerLoop(loop schedulerpb.SchedulerForFro
...
+		req.enqueue <- enqueueResult{status: failed}
+		return errors.New("scheduler is shutting down")
+
+	case schedulerpb.ERROR:
+		level.Warn(spanLogger).Log("msg", "scheduler returned error", "err", resp.Error)
+		req.enqueue <- enqueueResult{status: waitForResponse}
+		req.response <- &frontendv2pb.QueryResultRequest{
+			HttpResponse: &httpgrpc.HTTPResponse{
+				Code: http.StatusInternalServerError,
+				Body: []byte(resp.Error),
```
Comment (pstibrany) [inline pkg/frontend/v2/frontend_scheduler_worker.go:409]: Nice catch!

**Review by pstibrany — APPROVED** [review]
Nice, thank you!

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM

**Pull request merged**